pkg/apihelpers/apihelpers.go

```diff
@@ -324,6 +325,43 @@ func IsControllerConfigCompleted(ccName string, ccGetter func(string) (*mcfgv1.C
 	return fmt.Errorf("ControllerConfig has not completed: completed(%v) running(%v) failing(%v)", completed, running, failing)
 }
 
+// AreMCGeneratingSubControllersCompleted checks whether all MC producing sub-controllers are completed
+func AreMCGeneratingSubControllersCompleted(crcLister func(labels.Selector) ([]*mcfgv1.ContainerRuntimeConfig, error), mckLister func(labels.Selector) ([]*mcfgv1.KubeletConfig, error), selector labels.Selector) error {
+
+	containerConfigs, err := crcLister(selector)
+	if err != nil {
+		return err
+	}
+	for _, crc := range containerConfigs {
+		if crc.Generation != crc.Status.ObservedGeneration {
+			return fmt.Errorf("status for ContainerRuntimeConfig %s is being reported for %d, expecting it for %d", crc.ObjectMeta.Name, crc.Status.ObservedGeneration, crc.Generation)
+		}
+
+		for _, condition := range crc.Status.Conditions {
+			if condition.Type != mcfgv1.ContainerRuntimeConfigSuccess {
+				return fmt.Errorf("ContainerRuntimeConfig has not completed")
+			}
+		}
+	}
+
+	kubeletConfigs, err := mckLister(selector)
+	if err != nil {
+		return err
+	}
+	for _, mck := range kubeletConfigs {
+		if mck.Generation != mck.Status.ObservedGeneration {
+			return fmt.Errorf("status for KubeletConfig %s is being reported for %d, expecting it for %d", mck.ObjectMeta.Name, mck.Status.ObservedGeneration, mck.Generation)
+		}
+
+		for _, condition := range mck.Status.Conditions {
+			if condition.Type != mcfgv1.KubeletConfigSuccess {
```

Comment:
Doesn't affect this PR, but I'm curious if we actually reflect this up further, or is just ends up being logged in the controller :thinking:

Reply:
I remember seeing similar logs for the `IsControllerConfigComplete` in the controller